From 4e98563b269defde149ea6560a5d7d4f2487baa3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:48 -0400
Subject: [powerpc] eeh: EEH_PE_ISOLATED not reflect HW state

Message-id: <1410545655-205645-480-git-send-email-dzickus@redhat.com>
Patchwork-id: 94540
O-Subject: [RHEL7 PATCH 479/626] powerpc/eeh: EEH_PE_ISOLATED not reflect HW state
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 7b401850a1c9b64b010b3d107321df4e7d066995
Author: Gavin Shan <gwshan@linux.vnet.ibm.com>
Date:   Thu Apr 24 18:00:11 2014 +1000

    powerpc/eeh: EEH_PE_ISOLATED not reflect HW state

    When doing PE reset, EEH_PE_ISOLATED is cleared unconditionally.
    However, We should remove that if the PE reset has cleared the
    frozen state successfully. Otherwise, the flag should be kept.
    The patch fixes the issue.

    Signed-off-by: Gavin Shan <gwshan@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>

backport notes:  In my attempt to resolve in earlier commit conflict
I prematurely pulled in upstream commit 26833a5029b71, which deletes
half of this commit, hence the difference. :-/
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/eeh.c b/arch/powerpc/kernel/eeh.c
index d985e5d..f1a0bb5 100644
--- a/arch/powerpc/kernel/eeh.c
+++ b/arch/powerpc/kernel/eeh.c
@@ -628,8 +628,10 @@ int eeh_reset_pe(struct eeh_pe *pe)
   eeh_reset_pe_once(pe);
 
   rc = eeh_ops->wait_state(pe, PCI_BUS_RESET_WAIT_MSEC);
-  if ((rc & flags) == flags)
+  if ((rc & flags) == flags) {
+   eeh_pe_state_clear(pe, EEH_PE_ISOLATED);
    return 0;
+  }
 
   if (rc < 0) {
    pr_err("%s: Unrecoverable slot failure on PHB#%d-PE#%x",
-- 
1.7.1